### PR TITLE
[RAPTOR-5493] Validate output data for transforms in predict server

### PR DESCRIPTION
**Pull request by jmbannon**
# This repository is public. Do not put here any private DataRobot or customer's data: code, datasets, model artifacts, .etc.

## Summary
We never validated output data in the prediction server for transforms. This PR adds that support and a functional test.

## Changes
- Change abstract transform method to `_transform`, call it and add schema validation in `transform`
- Fix schema fail test to reach the assertion (would throw an assert error and get caught by `with pytest.raises`)

### CODE REVIEW

File: custom_model_runner/datarobot_drum/resource/predict_mixin.py [hunk 1]
```diff
@@ -182,32 +183,40 @@ def _transform(self, logger=None):
...
-            out_target = None
+            else:
+                out_data, _ = self._predictor.transform(
+                    binary_data=feature_binary_data,
+                    mimetype=feature_mimetype,
+                    charset=feature_charset,
+                )
+                out_target = None
+        except DrumSchemaValidationException as e:
+            response_status = HTTP_422_UNPROCESSABLE_ENTITY
```
Comment (jmbannon) [inline custom_model_runner/datarobot_drum/resource/predict_mixin.py:218]: Perhaps change this to its own error code?

**Review by scottp-ml — APPROVED** [review]
LGTM

File: tests/fixtures/validate_transform_fail_output_schema_validation/custom.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    class_order : A two element long list dictating the order of classes which should be used for
+        modeling. Class order will always be passed to fit by DataRobot for classification tasks,
+        and never otherwise. When models predict, they output a likelihood of one class, with a
+        value from 0 to 1. The likelihood of the other class is 1 - this likelihood. Class order
+        dictates that the first element in the list will be the 0 class, and the second will be the
+        1 class.
+    row_weights: An array of non-negative numeric values which can be used to dictate how important
+        a row is. Row weights is only optionally used, and there will be no filtering for which
+        custom models support this. There are two situations when values will be passed into
+        row_weights, during smart downsampling and when weights are explicitly provided by the user
```
Comment (4dahalibut) [inline tests/fixtures/validate_transform_fail_output_schema_validation/custom.py:38]: given this is a test fixture, a nit is just nuke all these comments. I don't think they're really needed for a fixture.

File: custom_model_runner/datarobot_drum/resource/predict_mixin.py [hunk 3]
```diff
@@ -182,32 +183,40 @@ def _transform(self, logger=None):
...
+                        target_charset,
+                    ) = self._fetch_data_from_request("y", logger=logger)
+                    mimetype_support_error_response = self._check_mimetype_support(target_mimetype)
+                    if mimetype_support_error_response is not None:
+                        return mimetype_support_error_response
+                except ValueError as e:
+                    response_status = HTTP_422_UNPROCESSABLE_ENTITY
+                    return {"message": "ERROR: " + str(e)}, response_status
+
+                out_data, out_target = self._predictor.transform(
```
Comment (4dahalibut) [inline custom_model_runner/datarobot_drum/resource/predict_mixin.py:201]: would you mind failing based on target type as we discussed?

Comment (4dahalibut) [reply custom_model_runner/datarobot_drum/resource/predict_mixin.py:201]: if Y is modified...

Comment (4dahalibut) [reply custom_model_runner/datarobot_drum/resource/predict_mixin.py:201]: nvm I see you created a ticket.

**Review by 4dahalibut — APPROVED** [review]

**Pull request merged**